resolver/src/main/java/io/netty/resolver/DefaultHostsFileEntriesResolver.java

```diff
@@ -30,33 +30,61 @@
  */
 public final class DefaultHostsFileEntriesResolver implements HostsFileEntriesResolver {
 
-    private final Map<String, Inet4Address> inet4Entries;
-    private final Map<String, Inet6Address> inet6Entries;
+    private final Map<String, List<InetAddress>> inet4Entries;
+    private final Map<String, List<InetAddress>> inet6Entries;
 
     public DefaultHostsFileEntriesResolver() {
         this(parseEntries());
     }
 
     // for testing purpose only
-    DefaultHostsFileEntriesResolver(HostsFileEntries entries) {
-        inet4Entries = entries.inet4Entries();
-        inet6Entries = entries.inet6Entries();
+    DefaultHostsFileEntriesResolver(HostsFileEntriesProvider entries) {
+        inet4Entries = entries.ipv4Entries();
+        inet6Entries = entries.ipv6Entries();
     }
 
     @Override
     public InetAddress address(String inetHost, ResolvedAddressTypes resolvedAddressTypes) {
+        String normalized = normalize(inetHost);
+        switch (resolvedAddressTypes) {
+            case IPV4_ONLY:
+                return firstAddress(inet4Entries.get(normalized));
+            case IPV6_ONLY:
+                return firstAddress(inet6Entries.get(normalized));
+            case IPV4_PREFERRED:
+                InetAddress inet4Address = firstAddress(inet4Entries.get(normalized));
+                return inet4Address != null ? inet4Address : firstAddress(inet6Entries.get(normalized));
+            case IPV6_PREFERRED:
+                InetAddress inet6Address = firstAddress(inet6Entries.get(normalized));
+                return inet6Address != null ? inet6Address : firstAddress(inet4Entries.get(normalized));
+            default:
+                throw new IllegalArgumentException("Unknown ResolvedAddressTypes " + resolvedAddressTypes);
+        }
+    }
+
+    /**
+     * Resolves all addresses of a hostname against the entries in a hosts file, depending on the specified
+     * {@link ResolvedAddressTypes}.
+     *
+     * @param inetHost the hostname to resolve
+     * @param resolvedAddressTypes the address types to resolve
+     * @return all matching addresses
```

Comment:
nit: either we should tell that this method returns `null` as well or return an empty List in this case.

Reply:
fixed